pennylane/qnn/cost.py

```diff
@@ -0,0 +1,119 @@
+            final step of each circuit
+        device (Device, Sequence[Device]): Corresponding device(s) where the resulting
+            function should be executed. This can either be a single device, or a list
+            of devices of length matching the number of observables.
+        interface (str, None): Which interface to use.
+            This affects the types of objects that can be passed to/returned to the function.
+            Supports all interfaces supported by the :func:`~.qnode` decorator.
+        diff_method (str, None): The method of differentiation to use with the created function.
+            Supports all differentiation methods supported by the :func:`~.qnode` decorator.
+
+    Returns:
+        callable: a loss function with signature ``loss(*args, target=None, **kwargs)`` that calculates
+        the squared error loss between the observables' expectation values and a target.
+
+    .. seealso:: :func:`~.map`
+
+    **Example:**
+
+    First, we create a device and design an ansatz:
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=3)
+
+        def ansatz(phis, **kwargs):
+            for w, phi in enumerate(phis):
+                qml.RX(phi, wires=w)
+
+    Now we can create the observables:
+
+    .. code-block:: python3
+
+        obs = [
+            qml.PauliZ(0),
+            qml.PauliX(0),
+            qml.PauliZ(1) @ qml.PauliZ(2)
+        ]
+
+    Next, we can define the loss function:
+
+    >>> loss = qml.qnn.SquaredErrorLoss(ansatz, observables, dev, interface="torch")
+    >>> phis = np.ones(num_qubits)
+    >>> loss(phis, target=np.array([1.0, 0.5, 0.1]))
+    tensor([0.2113, 0.2500, 0.0368], dtype=torch.float64)
+
+    The loss function can be minimized using any gradient descent-based
+    :doc:`optimizer </introduction/optimizers>`.
+    """
+
+    def __init__(
+        self,
+        ansatz,
+        observables,
+        device,
+        measure="expval",
+        interface="autograd",
+        diff_method="best",
+        **kwargs
+    ):
+        self.qnodes = qml.map(
+            ansatz,
+            observables,
+            device,
+            measure=measure,
+            interface=interface,
+            diff_method=diff_method,
+            **kwargs
+        )
+
+    def loss(self, *args, target=None, **kwargs):
+        r"""Calculates the squared error loss between the observables'
+        expectation values and a target.
+
+        Keyword Args:
+            target (tensor): target values
+
+        Returns:
+            array[float]: squared error values
+        """
+
+        if target is None:
+            raise ValueError("The target cannot be None")
+
+        input_ = self.qnodes(*args, **kwargs)
+
+        if len(target) != len(input_):
+            raise ValueError(
+                "Input target of incorrect length %d instead of %d" % (len(target), len(input_))
```

Comment:
Might be better to use an fstring here, e.g.,
```suggestion
                f"Input target of incorrect length {len(target)} instead of {len(input_)}"
```
(I think that's the right syntax 🤔

Reply:
Yeah, I had a doubt about compatibility, but then I realized that Pennylane explicitly requires Python >= 3.6... so we're good to go!